system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "subscriber_topic_statistics.hpp"
+#include "system_metrics_collector/constants.hpp"
+#include "system_metrics_collector/metrics_message_publisher.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+using std::placeholders::_1;
+
+namespace topic_statistics_collector
+{
+template<typename T>
+SubscriberTopicStatistics<T>::SubscriberTopicStatistics(
+  const std::string & node_name,
+  const rclcpp::NodeOptions & node_options,
+  const std::string & topic_name)
+: rclcpp_lifecycle::LifecycleNode{node_name, node_options},
+  topic_name_(topic_name)
+{
```

Comment:
This constructor is very complex, let's define helper functions (free function, static, CPP only, just for readability purposes). I'd do as follow:

 * `declareParameters`
  * `buildParameterDescriptor`
    * `buildIntegerRange`
 * `initializeCollectors`
 * `initializeSubscriptions`